src/Sarif/Writers/ResultLogJsonWriter.cs

```diff
@@ -167,14 +216,36 @@ public void Dispose()
                 _jsonWriter.WriteEndObject(); // End: resultsLog
             }
 
-            _writeState = State.Disposed;
+            _writeConditions |= Conditions.Disposed;
+        }
+
+        private void EnsureInitialized()
+        {
+            if (_writeConditions == Conditions.None)
+            {
+                Initialize();
+            }
+        }
+
+        private void EnsureStateNotAlreadySet(Conditions invalidConditions)
+        {
+            Conditions observedInvalidConditions = _writeConditions & invalidConditions;
+            if (observedInvalidConditions != Conditions.None)
+            {
+                // 	InvalidState	One or more invalid states were detected during serialization: {0}	
+                throw new InvalidOperationException(string.Format(SarifResources.InvalidState, observedInvalidConditions));
```

Comment:
Ok, now I know what it means. I think we need to be much friendlier. (I'll come back to this comment in a minute...)